demo/src/main/scala/demo/DemoMain.scala

```diff
@@ -13,6 +14,158 @@ import scala.scalajs.js.typedarray.{ArrayBuffer, Uint8Array}
+          case Parsed.Failure(lastParser, index, extra) =>
+            metadataBox.textContent = "Failed to parse midi\n" + extra.traced.trace
+        }
+      }
+    }
+
+  }
+  val notes = Vector("C","C#","D","D#","E","F","F#","G","G#","A","A#","B")
+  lazy val allNotes = for(octave <- 0 to 10; note <- notes) yield (note, octave)
+  def makeNote(note: String, octave: Int) = {
+    js.Dynamic.newInstance(js.Dynamic.global.Audio)(
+      js.Dynamic.global.Synth.generate(0, note, octave - 2, 10)
+    )
+  }
+  def playMusic(midi: fastparse.Midi,
+                noteLogger: String => Unit,
+                progressLogger: String => Unit) = {
+    import fastparse.Midi._, MidiData._
+    case class TrackInfo(var savedTicks: Int, var track: List[(Int, Midi.TrackEvent)]){
+      def tillNext = track.head._1 - savedTicks
+    }
+    noteLogger("Initializing Musical Notes...")
+    val notes =
+      midi.tracks
+          .flatten
+          .collect{case (_, MidiEvent(channel, NoteOn(note, velocity))) => (channel, note)}
+          .distinct
+          .map{ case (channel, note) => ((channel, note), makeNote(allNotes(note)._1, allNotes(note)._2)) }
+          .toMap
+
+    val allTracks = midi.tracks.map(_.toList).map(TrackInfo(0, _))
+    // 120bpm, in microseconds-per-quarter-beat
+    var currentTempo = 1000000 * 120 / 60 / 4
+
+    def tick(): Unit = {
+
+      val logs = collection.mutable.Buffer.empty[String]
+      progressLogger(allTracks.map(_.track.length).mkString(", "))
+      while ({
+        val remaining = allTracks.filter(_.track != Nil)
+        val nextTrack = remaining.minBy(_.tillNext)
+        val tillNext = nextTrack.tillNext
+
+        if (tillNext == 0) {
+          val nextEvent = nextTrack.track.head._2
+          nextTrack.track = nextTrack.track.tail
+          nextTrack.savedTicks = 0
+          logs.append(nextEvent.toString)
+          nextEvent match{
+            case MetaEvent.Tempo(n) =>
+              currentTempo = n
+
+            case MidiEvent(channel, midiData) =>
+              midiData match{
+                case NoteOn(note, velocity) =>
+                  notes((channel, note)).volume = velocity * 1.0 / 127
+                  notes((channel, note)).play()
+                case NoteOff(note, velocity) =>
+                  notes((channel, note)).pause()
+                  notes((channel, note)).currentTime = 0
+                case _ => // skip the rest
+              }
+
+            case _ => // skip the rest
+          }
+          true
+        } else if (remaining.nonEmpty){
+
+          val milliSleep = midi.tickDiv match{
+            case fastparse.Midi.TickDiv.Metric(divisionsPerQuarterBeat) =>
+              val milliTempo = currentTempo / 1000
+              milliTempo * tillNext / divisionsPerQuarterBeat
+            case fastparse.Midi.TickDiv.TimeCode(fps, divs) =>
+              ???
```

Comment:
Let's leave it `???` for now. I couldn't find any midi files on my machine that would trigger that code path, so I'm happy to leave it as "not implemented" for the time being. If someone complains about it not working, then I'll look into fixing it